lib/models/asyncapi.js

```diff
@@ -276,121 +351,139 @@ function addNameToKey(messages, number) {
+ * @param {SchemaTypesToIterate[]} schemaTypesToIterate
  */
-function crawl(schema, seenObj, callback) {
-  if (isCircular(schema, seenObj)) return true;
+function traverseSchema(schema, cb, prop, schemaTypesToIterate) {
+  if (schema === null) return;
+  if (!schemaTypesToIterate.includes(SchemaTypesToIterate.arrays) && schema.type() === 'array') return;
+  if (!schemaTypesToIterate.includes(SchemaTypesToIterate.objects) && schema.type() === 'object') return;
+  if (schema.isCircular()) return;
+  if (cb(schema, prop, SchemaIteratorCallbackType.NEW_SCHEMA) === false) return;
 
-  seenObj.push(schema.json());
-  callback(schema);
   if (schema.type() !== undefined) {
     switch (schema.type()) {
     case 'object':
-      recursiveSchemaObject(schema, seenObj, callback);
+      recursiveSchemaObject(schema, cb, schemaTypesToIterate);
       break;
     case 'array':
-      recursiveSchemaArray(schema, seenObj, callback);
+      recursiveSchemaArray(schema, cb, schemaTypesToIterate);
       break;
     }
   } else {
+    traverseCombinedSchemas(schema, cb, schemaTypesToIterate);
+  }
+  cb(schema, prop, SchemaIteratorCallbackType.END_SCHEMA);
+}
+
+/**
+ * Traverse combined notions
+ * 
+ * @private
+ * @param {Schema} schemaContent schema.
+ * @param {TraverseSchemas} cb 
+ * @param {SchemaTypesToIterate[]} schemaTypesToIterate
+ */
+function traverseCombinedSchemas(schema, cb, schemaTypesToIterate) {
   //check for allOf, oneOf, anyOf
-    const checkCombiningSchemas = (combineArray) => {
-      if (combineArray !== null && combineArray.length > 0) {
-        combineArray.forEach(combineSchema => {
-          if (crawl(combineSchema, seenObj, callback)) markCircular(schema);
-        });
-      }
-    };
+  const checkCombiningSchemas = (combineArray) => {
+    if (combineArray !== null && combineArray.length > 0) {
+      combineArray.forEach(combineSchema => {
+        traverseSchema(combineSchema, cb, null, schemaTypesToIterate);
+      });
+    }
+  };
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.allOfs)) {
     checkCombiningSchemas(schema.allOf());
+  }
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.anyOfs)) {
     checkCombiningSchemas(schema.anyOf());
+  }
+  if (schemaTypesToIterate.includes(SchemaTypesToIterate.oneOfs)) {
     checkCombiningSchemas(schema.oneOf());
   }
-
-  seenObj.pop();
 }
 
 /**
- * Go through each channel and for each parameter, and message payload and headers recursively call the callback for each schema.
+ * Go through each channel and for each parameter, and message payload and headers recursively call the cb for each schema.
  * 
  * @private
  * @param {AsyncAPIDocument} doc 
- * @param {FoundSchemaCallback} callback
+ * @param {FoundSchemaCallback} cb
+ * @param {SchemaTypesToIterate[]} schemaTypesToIterate
  */
-function schemaDocument(doc, callback) {
+function traverseAsyncApiDocument(doc, cb, schemaTypesToIterate) {
+  if (!schemaTypesToIterate) {
+    schemaTypesToIterate = [
```

Comment:
We can simplify it by:

```js
schemaTypesToIterate = Object.values(SchemaTypesToIterate)
```

:)

Reply:
I remember trying different solutions with for loops but yes, this is way better :smile: